### PR TITLE
CMake unity and nonunity builds per project (RIPD-1326):

### PR DESCRIPTION
**Pull request by ximinez**
* Still respects "-Dtarget" unity/nonunity selection. Still defaults to
  unity.
* Adds a new target (rippled_classic or rippled_unity) depending on
  unity/nonunity selection.
* New target does not build by default.
  eg. Use `cmake --build . -target rippled_classic`
* Copy all config variables (Debug/Release to DebugClassic/ReleaseClassic) for nonunity builds
* CI uses the more generic "cmake --build" command

...I noticed and fixed at the last minute some indentation that got borked. Please keep an eye out for any I may have missed.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -99,12 +99,19 @@ setup_build_cache()
...
-  set(CMAKE_EXE_LINKER_FLAGS_RELEASECLASSIC
-    ${CMAKE_EXE_LINKER_FLAGS_RELEASE})
+    get_cmake_property(allvars VARIABLES)
+    string(REGEX MATCHALL "[^;]*(DEBUG|RELEASE)[^;]*" matchvars "${allvars}")
+    foreach(var IN LISTS matchvars)
+        string(REGEX REPLACE "(DEBUG|RELEASE)" "\\1CLASSIC" newvar ${var})
+        set(${newvar} ${${var}})
+    endforeach()
+
+    get_cmake_property(allvars CACHE_VARIABLES)
```
Comment (bachase) [inline CMakeLists.txt]: What is the difference between `CACHE_VARIABLES` and `VARIABLES`?

File: CMakeLists.txt [hunk 2]
```diff
@@ -409,23 +411,40 @@ if (NOT is_msvc)
...
 
+list(APPEND rippled_src_unity "${rippled_src_all}")
+list(APPEND rippled_src_nonunity "${rippled_src_all}")
+
 ############################################################
 
 if (WIN32 OR is_xcode)
   group_sources(src)
   group_sources(docs)
+  group_sources(Builds)
```
Comment (bachase) [inline CMakeLists.txt]: How come Builds is being added?  It looks like it adds `CMakeFuncs.cmake` to the IDE?

File: CMakeLists.txt [hunk 3]
```diff
@@ -157,173 +164,174 @@ set(soci_extra_includes
...
+${rocks_db_system_header})
+
+add_with_props(rippled_src_unity src/ripple/unity/soci_ripple.cpp ${soci_extra_includes})
+
+list(APPEND ripple_unity_srcs ${beast_unity_srcs} ${test_unity_srcs}
+src/ripple/unity/nodestore.cpp
+src/ripple/unity/soci_ripple.cpp
+src/unity/nodestore_test_unity.cpp)
+
+#  set_property(
```
Comment (bachase) [inline CMakeLists.txt]: Drop the commented out section?

File: CMakeLists.txt (hunk 1 above)
Comment (ximinez) [inline CMakeLists.txt]: Normal variables are basically just local, where cache variables persist between runs in `CMakeCache.txt`. I wanted to make sure to get both.

* https://cmake.org/cmake/help/v3.4/command/set.html
* https://cmake.org/cmake/help/v3.4/command/get_cmake_property.html
* https://cmake.org/cmake/help/v3.4/prop_dir/CACHE_VARIABLES.html
* https://cmake.org/cmake/help/v3.4/prop_dir/VARIABLES.html

File: CMakeLists.txt (hunk 3 above)
Comment (ximinez) [inline CMakeLists.txt]: Woops. Yeah. Will fix.

**Comment by ximinez** [comment]
@bachase Yeah, there are some more changes in the pipeline, and I'm using the fact that it's still experimental as an excuse to hold off writing docs.

File: CMakeLists.txt [hunk 4]
```diff
@@ -157,173 +164,159 @@ set(soci_extra_includes
...
+        file(GLOB_RECURSE cursrcs src/ripple/${curdir}/*.cpp)
+        list(APPEND rippled_headers "${cursrcs}")
+    endforeach()
+    list(APPEND rippled_src_nonunity "${rippled_headers}")
+
+    set_property(
+        SOURCE ${rippled_headers}
+        APPEND
+        PROPERTY HEADER_FILE_ONLY
+        true)
```
Comment (seelabs) [inline CMakeLists.txt]: Doesn't this unconditionally set non-unity source files as header files, even for non-unity builds? I can't (easily) test with an IDE, but that does the right thing in visual studio when the current configuration is a classic build?

Comment (seelabs) [inline CMakeLists.txt]: Oops, I did misread the code. Nevermind.

**Comment by seelabs** [comment]
This looks good. :+1:

**Comment by nbougalis** [comment]
Merged as 361917e902f7ba16462ac7b4d8a16f9da4bc55d6

**Pull request closed without merging**